### PR TITLE
repository

### PR DESCRIPTION
**Pull request by shayb**
update local and remote data

### CODE REVIEW

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/presenter/AppsDiscoveryListPresenter.kt [hunk 1]
```diff
@@ -1,35 +1,16 @@
...
+        view?.updateData(apps)
     }
 
     override fun onAppClicked(app: EcosystemApp) {
         view?.navigateToApp(app)
     }
-
-    override fun onAttach(view: IAppsDiscoveryListView) {
-        super.onAttach(view)
-        loadData()
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/presenter/AppsDiscoveryListPresenter.kt:31]: check that it's on the presenter logic

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/view/AppsDiscoveryActivity.kt [hunk 2]
```diff
@@ -5,16 +5,46 @@ import android.content.Intent
...
+        findViewById<Button>(R.id.get).setOnClickListener({
+            getData()
+        })
+//        findViewById<Button>(R.id.clear).setOnClickListener({
+//            repository?.clearLocalData()
+//            findViewById<AppsDiscoveryList>(R.id.list).updateData(listOf())
+//        })
+    }
+
+    private fun getData() {
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/view/AppsDiscoveryActivity.kt]: move to the presenter logic

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/view/AppsDiscoveryActivity.kt [hunk 3]
```diff
@@ -5,16 +5,46 @@ import android.content.Intent
...
+        repository?.getDiscoveryApps(object : OperationResultCallback<List<EcosystemApp>?>{
+            override fun onResult(result: List<EcosystemApp>?) {
+                result?.let {
+                    findViewById<AppsDiscoveryList>(R.id.list).updateData(result)
+                }
+            }
+
+            override fun onError(error: String) {
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/view/AppsDiscoveryActivity.kt]: consider changing to some General module exception

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/view/AppsDiscoveryActivity.kt [hunk 4]
```diff
@@ -5,16 +5,46 @@ import android.content.Intent
...
+                //TODO
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/view/AppsDiscoveryActivity.kt]: remove? or do something

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/view/AppsDiscoveryList.kt [hunk 5]
```diff
@@ -99,7 +99,6 @@ class AppsDiscoveryAdapter(private val presenter: IAppsDiscoveryListPresenter) :
                 setLineSpacing(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, app.fontLineSpacing,  context.resources.displayMetrics), 1.0f)
                 text = app.cardTitle
                 setTextSize(TypedValue.COMPLEX_UNIT_SP, app.cardFontSize)
-                   //TODO not working
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/view/AppsDiscoveryList.kt:102]: 🚀

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/model/EcosystemAppList.kt [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package org.kinecosystem.appsdiscovery.sender.model
+
+import android.os.Parcelable
+import com.google.gson.annotations.SerializedName
+import kotlinx.android.parcel.Parcelize
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/model/EcosystemAppList.kt:5]: how much it's adding to the module aar?

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/model/EcosystemAppResponse.kt [hunk 7]
```diff
@@ -8,7 +8,9 @@ import kotlinx.android.parcel.Parcelize
 @Parcelize
 data class EcosystemAppResponse(
         @SerializedName("version")
-        val version: String?,
+        val version: String="",
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/model/EcosystemAppResponse.kt]: format

Comment (OrenZak) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/model/EcosystemAppResponse.kt]: String comparison is more expensive than int

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsLocal.kt [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import com.google.gson.Gson
+import com.google.gson.JsonSyntaxException
+import com.google.gson.reflect.TypeToken
+import org.kinecosystem.appsdiscovery.base.LocalStore
+import org.kinecosystem.appsdiscovery.sender.model.EcosystemApp
+
+
+private const val APPS_DISCOVERY_STORE = "APPS_DISCOVERY_STORE"
+private const val APPS_DISCOVERY_VERSION_KEY = "APPS_DISCOVERY_VERSION_KEY"
+private const val APPS_DISCOVERY_KEY = "APPS_DISCOVERY_KEY"
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsLocal.kt]: can you use companion object?

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt [hunk 9]
```diff
@@ -9,13 +8,15 @@ import org.kinecosystem.appsdiscovery.sender.model.EcosystemAppResponse
 import java.io.IOException
 import java.util.concurrent.TimeUnit
 
+
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt]: move the below to companion object

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt [hunk 10]
```diff
@@ -28,30 +29,28 @@ class Network {
         httpClient = httpClientBuilder.build()
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt:33]: you can use `with(httpClientBuilder)`

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt [hunk 11]
```diff
@@ -28,30 +29,28 @@ class Network {
...
-    private fun getRequest(url: String): Request = Request.Builder().url(url).build()
-
-    fun getDiscoveryApps() {
-        httpClient.newCall(getRequest(GET_DISCOVERY_APPS_STAGE_URL)).enqueue(object : Callback {
+    fun getDiscoveryAppsServerData(callback: OperationResultCallback<EcosystemAppResponse>) {
+        httpClient.newCall(getRequest(GET_DISCOVERY_APPS_PROD_URL)).enqueue(object : Callback {
             override fun onFailure(call: Call, e: IOException) {
-                //TODO
-                Log.d("", "response failed")
+                callback.onError("server failed get discoveryApps ${e.message}")
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt:39]: Pass here the actual exception, and the repo should create a ModuleException with some logic.

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt [hunk 12]
```diff
@@ -28,30 +29,28 @@ class Network {
...
             }
 
             override fun onResponse(call: Call, response: Response) {
                 if (response.isSuccessful) {
                     try {
                         val responseData = gson.fromJson(response.body()?.charStream(), EcosystemAppResponse::class.java)
-                        Log.d("", "####response version  ${responseData.version}")
-                        Log.d("", "##### app size ${responseData.apps?.size} app  ${responseData.apps?.get(0)}")
+                        callback.onResult(responseData)
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt:46]: check body and check the char stream why is that.

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRepository.kt [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+import android.os.Handler
+import android.os.Looper
+import org.kinecosystem.appsdiscovery.sender.model.EcosystemApp
+import org.kinecosystem.appsdiscovery.sender.model.EcosystemAppResponse
+import org.kinecosystem.appsdiscovery.sender.model.hasNewData
+
+class DiscoveryAppsRepository(context: Context){
+
+    private var local = DiscoveryAppsLocal(context)
+    private var remote = DiscoveryAppsRemote()
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRepository.kt]: `local`, `remote`  and `handler` vars should be on ctr as param for unit testing usage.

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRepository.kt [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    private var hasLocalData = false
+    private val handler =  Handler(Looper.getMainLooper())
+
+    fun getDiscoveryApps(discoveryAppsCallback: OperationResultCallback<List<EcosystemApp>?>) {
```
Comment (OrenZak) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRepository.kt]: should be observer, as you trigger twice.

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/presenter/AppsDiscoveryListPresenter.kt (hunk 1 above)
Comment (shayb) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/presenter/AppsDiscoveryListPresenter.kt:31]: changed the logic of loading is in the presenter

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/view/AppsDiscoveryActivity.kt (hunk 2 above)
Comment (shayb) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/view/AppsDiscoveryActivity.kt]: moved

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/model/EcosystemAppResponse.kt (hunk 7 above)
Comment (shayb) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/model/EcosystemAppResponse.kt]: cool

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsLocal.kt (hunk 8 above)
Comment (shayb) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsLocal.kt]: cool

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt (hunk 9 above)
Comment (shayb) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt]: cool

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt (hunk 10 above)
Comment (shayb) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt:33]: cool

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt (hunk 12 above)
Comment (shayb) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRemote.kt:46]: changed to string

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRepository.kt (hunk 13 above)
Comment (shayb) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/repositories/DiscoveryAppsRepository.kt]: cool

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/base/LocalStore.kt [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+
+class LocalStore(context: Context, name:String){
+
+    private val sharedPreferences:SharedPreferences = context.applicationContext.getSharedPreferences(name, Context.MODE_PRIVATE)
+
+    fun updateString(key:String, value:String){
+        sharedPreferences.edit().apply{
+            putString(key, value)
+            apply()
+        }
```
Comment (berryve) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/base/LocalStore.kt:14]: if you use android ktx you can minimize the code to:
```
sharedPreferences.edit {
    putString("key", value)
}
```

See [https://developer.android.com/kotlin/ktx#kotlin](https://developer.android.com/kotlin/ktx#kotlin)

And then maybe the LocalStore is not needed at all ...

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/presenter/AppsDiscoveryListPresenter.kt [hunk 16]
```diff
@@ -1,35 +1,39 @@
...
-//            override fun onResult(apps: List<EcosystemApp>?) {
-//                apps?.let {
-//                    view?.updateData(apps)
-//                }
-//            }
-//
-//            override fun onError(errorCode: Int) {
-//            }
-//        })
+class AppsDiscoveryListPresenter(private val discoveryAppsRepository: DiscoveryAppsRepository) : BasePresenter<IAppsDiscoveryListView>(), IAppsDiscoveryListPresenter, Observer {
```
Comment (berryve) [inline appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/presenter/AppsDiscoveryListPresenter.kt:9]: I thought we were not using 'I' as naming convention for interfaces

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/base/LocalStore.kt (hunk 15 above)
Comment (shayb) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/base/LocalStore.kt:14]: we are thinking or removing the ktx lib

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/presenter/AppsDiscoveryListPresenter.kt (hunk 16 above)
Comment (shayb) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/discovery/presenter/AppsDiscoveryListPresenter.kt:9]: as i see in the core it has

File: appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/model/EcosystemAppList.kt (hunk 6 above)
Comment (shayb) [reply appsDiscovery/src/main/java/org/kinecosystem/appsdiscovery/sender/model/EcosystemAppList.kt:5]: probably will be removed later on

**Pull request merged**